include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h

```diff
@@ -538,25 +511,28 @@ struct __get_sycl_range
     auto
     operator()(_Iter __first, _Iter __last)
         -> ::std::enable_if_t<is_temp_buff<_Iter>::value && !is_zip<_Iter>::value && !is_permutation<_Iter>::value,
-                              __buffer_holder<val_t<_Iter>, AccMode>>
+                              __range_holder<oneapi::dpl::__ranges::all_view<val_t<_Iter>, AccMode>>>
     {
         static_assert(!oneapi::dpl::__internal::is_const_iterator<_Iter>::value || AccMode == sycl::access::mode::read,
                       "Should be non-const iterator for a modifying algorithm.");
 
         assert(__first < __last);
 
-        using copy_direct_tag = ::std::integral_constant<bool, AccMode == sycl::access::mode::read_write ||
-                                                                   AccMode == sycl::access::mode::read>;
-        using copy_back_tag = ::std::integral_constant<bool, AccMode == sycl::access::mode::read_write ||
-                                                                 AccMode == sycl::access::mode::write>;
+        using _T = val_t<_Iter>;
+
+        const bool __is_copy_direct = AccMode == sycl::access::mode::read_write || AccMode == sycl::access::mode::read;
+        const bool __is_copy_back = AccMode == sycl::access::mode::read_write || AccMode == sycl::access::mode::write;
 
-        auto buf = copy_direct(__first, __last, copy_direct_tag());
-        buf = copy_back(__first, buf, copy_back_tag());
+        auto buf = __is_copy_direct ? sycl::buffer<_T, 1>(std::addressof(*__first), __last - __first)
+                                    : sycl::buffer<_T, 1>(__last - __first);
+        buf.set_write_back(__is_copy_back);
 
+        //a temp buffer lifetime extension
         using BufferType = oneapi::dpl::__internal::__lifetime_keeper<decltype(buf)>;
         m_buffers.push_back(::std::make_shared<BufferType>(buf));
 
-        return __buffer_holder<val_t<_Iter>, AccMode>{buf};
+        return __range_holder<oneapi::dpl::__ranges::all_view<_T, AccMode>>{
+            oneapi::dpl::__ranges::all_view<_T, AccMode>(buf)};
```

Comment:
Actually, a buffer lifetime extension via `__buffer_holder` was redundant, because we have already extended the lifetime in 
`m_buffers.push_back(::std::make_shared<BufferType>(buf)); `line.

Reply:
Well, that relies on the `__get_sycl_range` object staying in scope, which I don't think will be the case for the async APIs, or am I missing something?

Reply:
Now that I think about it, if the `sycl::accessor` is enough to keep the buffer alive, do we gain anything from having `m_buffers` as a member variable in the first place?  Since the scope of the `struct __get_sycl_range` likely does not survive beyond the accessor its `operator()` returns, perhaps we don't gain anything from it.  

Of course, we do need to be sure that the `sycl::accessor` does in fact keep the buffer alive to consider this.

Reply:
Regarding  "the async APIs" - on one of "Pass data..." tech meetings we came to conclusion that we will no support such case - "host data + Async oneDPL call", just USM or hetero iterators for async cases.

Reply:
Aha, OK I understand the situation fully now, thanks for the clarification.

On the other hand ... Would we be better off to remove the `m_buffers` member variable, and instead keep `struct __buffer_holder` return to accomplish the lifetime extension?  

With `__get_sycl_range` it is quite unclear that this must be kept in scope for the lifetime of all kernels using the data.   However, what the `operator()` returns should stay in scope as it is what we use to represent the data.  In practice, perhaps they are the same because almost surely people will capture the return with `auto` within the scope of the `__get_sycl_range` object.  Switching is not a required change, but something to consider.

Reply:
"instead keep struct __buffer_holder".
For simple cases - yes... and there was such approach in the first version of the code.
But there are more complicated cases zip_iterator of zip_iterators of host iterators and hetero iteratos. In result we have got some hierarchical type - range. That range will be pass  to the SYCL backend and futher to kernel. We cannot keep sycl::buffer on that range. So, it range is based on placeholder accessors. For "flat" zip_iterator with may keep sycl::buffer(s) within  `__get_sycl_range`  lifetime as tuple f.e., but for hierarchical zip it is so complicated to do it compile time. As simpler solution  - was introduced storing buffers in RT via vector of smart pointers.

Reply:
I don't think I understand.  Prior to this PR, we return a `__buffer_holder` from the host_iterator specialization of `operator()`, and a `__range_holder` from the other specializations.  Why cant we just just continue to do that, and then remove `m_buffer`, which is only currently used for the host_iterator specialization?

To me it seems that it is just two ways of removing a redundancy.  Again, I think it is preference, but I want to understand why that option may not be feasible.

Reply:
We cannot "just continue" because we have lost sycl::buffer (__buffer_holder) life on this line (in case of zip):
`return oneapi::dpl::__ranges::make_zip_view(::std::get<_Ip>(tmp).all_view()...); `
https://github.com/oneapi-src/oneDPL/blob/main/include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h#L397

On this we have got __buffer_holder "on the fly", just in "a call scope" and immediately get an accessor (`oneapi::dpl::__ranges::all_view`)  by calling `all_view()`.

Reply:
Alright, I think I understand now.  Thanks for the continued explanation.  

To restate to make sure I understand:  we are using this same `__get_sycl_range` object, to recursively create a `__buffer_holder` and then generate the views of the wrapped iterators.  The views live-on in the returned fancy iterator `__range_holder`, but the created `__buffer_holder` for each wrapped iterator gets lost as it leaves scope in the helper function  `gen_zip_view` (as an example with `zip_iterator`).

Alright this makes sense.  I think the way you have it is best.